util/list.c

```diff
@@ -0,0 +1,744 @@
+{
+	struct list_desc *list = list_desc;
+	struct list_elem *prev;
+	struct list_elem *next;
+	struct list_elem *elem;
+
+	if (!list)
+		return FAILURE;
+
+	prev = NULL;
+	next = list->first;
+
+	elem = create_element(data, prev, next);
+	if (!elem)
+		return FAILURE;
+
+	update_links(prev, elem, next);
+
+	update_desc(list, elem, list->last);
+
+	list->nb_elements++;
+
+	return SUCCESS;
+}
+
+int32_t list_add_last(List_desc list_desc, void *data)
+{
+	struct list_desc *list = list_desc;
+	struct list_elem *prev;
+	struct list_elem *next;
+	struct list_elem *elem;
+
+	if (!list)
+		return FAILURE;
+
+	prev = list->last;
+	next = NULL;
+
+	elem = create_element(data, prev, next);
+	if (!elem)
+		return FAILURE;
+
+	update_links(prev, elem, next);
+
+	update_desc(list, list->first, elem);
+
+	list->nb_elements++;
+
+	return SUCCESS;
+}
+
+int32_t list_add_idx(List_desc list_desc, void *data, uint32_t idx)
+{
+	struct list_desc	*list = list_desc;
+
+	if (!list)
+		return FAILURE;
+
+	/* If there are no elements the creation of an iterator will fail */
+	if (list->nb_elements == 0 || idx == 0)
+		return list_add_first(list_desc, data);
+	if (list->nb_elements == idx)
+		return list_add_last(list_desc, data);
+
+	list->l_it.elem = list->first;
+	if (SUCCESS != iterator_move(&(list->l_it), idx))
+		return FAILURE;
+
+	return iterator_insert(&(list->l_it), data, 0);
+}
+
+int32_t list_add_find(List_desc list_desc, void *data)
+{
+	struct list_desc	*list = list_desc;
+	struct list_elem	*elem;
+
+	if (!list)
+		return FAILURE;
+
+	/* Based on place iterator */
+	elem = list->first;
+	while (elem) {
+		if (0 > list->comparator(data, elem->data))
+			break;
+		elem = elem->next;
+	}
+	if (elem == NULL) {
+		list->l_it.elem = list->last;
+		return iterator_insert(&(list->l_it), data, 1);
+	} else {
+		list->l_it.elem = elem;
+		return iterator_insert(&(list->l_it), data, 0);
+	}
+
+}
+
+int32_t list_edit_first(List_desc list_desc, void *new_data)
+{
+	struct list_desc *list = list_desc;
+
+	if (!list)
+		return FAILURE;
+
+	list->first->data = new_data;
+
+	return SUCCESS;
+}
+
+int32_t list_edit_last(List_desc list_desc, void *new_data)
+{
+	struct list_desc *list = list_desc;
+
+	if (!list)
+		return FAILURE;
+
+	list->last->data = new_data;
+
+	return SUCCESS;
+}
+
+int32_t list_edit_idx(List_desc list_desc, void *new_data, uint32_t idx)
+{
+	struct list_desc	*list = list_desc;
+
+	if (!list)
+		return FAILURE;
+
+	list->l_it.elem = list->first;
+	if (SUCCESS != iterator_move(&(list->l_it), idx))
+		return FAILURE;
+
+	return iterator_edit(&(list->l_it), new_data);
+}
+
+int32_t list_edit_find(List_desc list_desc, void *new_data, void *cmp_data)
+{
+	struct list_desc	*list = list_desc;
+
+	if (!list)
+		return FAILURE;
+
+	list->l_it.elem = list->first;
+	if (SUCCESS != iterator_find(&(list->l_it), cmp_data))
+		return FAILURE;
+
+	return iterator_edit(&(list->l_it), new_data);
+}
+
+void*	list_read_first(List_desc list_desc, int32_t *result)
```

Comment:
usually there is space between return type and function name not tab

Reply:
![image](https://user-images.githubusercontent.com/16192750/75542817-33f3dc00-5a29-11ea-8bc5-d2911377c20b.png)
When you collapse the functions or in the header file this way is more readable.
Should I replaece all with space?